File: src/main/java/dark/leech/text/plugin/PluginUpdate.java
package dark.leech.text.plugin;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import javax.swing.*;

import org.apache.commons.collections4.CollectionUtils;
import org.json.JSONArray;
import org.json.JSONObject;

import com.google.gson.Gson;

import dark.leech.text.action.Log;
import dark.leech.text.enities.PluginEntity;
import dark.leech.text.enities.RepositoryEntity;
import dark.leech.text.models.Repository;
import dark.leech.text.ui.notification.Toast;
import dark.leech.text.util.AppUtils;
import dark.leech.text.util.FileUtils;
import dark.leech.text.util.Http;
import dark.leech.text.util.SettingUtils;

public class PluginUpdate {

    private static final Gson gson = new Gson();

    private static PluginUpdate pluginUpdate;

    private PluginUpdate() {}

    public static PluginUpdate getUpdate() {
        if (pluginUpdate == null) pluginUpdate = new PluginUpdate();
        return pluginUpdate;
    }

    public void checkUpdate() {
        var repos = RepositoryManager.getManager().repositoryList();
        if (repos == null || repos.isEmpty()) return;

        int poolSize = Math.max(1, SettingUtils.MAX_CONN);
        ExecutorService executor = Executors.newFixedThreadPool(poolSize);
        for (RepositoryEntity repo : repos) {
            if (repo == null || !repo.isEnabled()) continue;
            executor.submit(
                    () -> {
                        try {
                            Log.add(
                                    "Checking update for repository: "
                                            + repo.getLink()
                                            + "  thread: "
                                            + Thread.currentThread().getName());
                            checkUpdate(repo.getLink());
                        } catch (Exception e) {
                            Log.add(e);
                        }
                    });
        }
        executor.shutdown();
    }

    public void checkUpdate(String repositoryLink) {
        Log.add("[PluginUpdate] Checking update from: " + repositoryLink);

        try {
            String js = Http.request(repositoryLink).string();
            if (js == null || js.isEmpty()) {
                Log.add("[PluginUpdate] Empty or null response from repository");
                return;
            }
            Log.add("[PluginUpdate] Repository response received, length: " + js.length());

            // Try vBook Repository format first
            Repository repository = gson.fromJson(js, Repository.class);
            if (repository != null && !CollectionUtils.isEmpty(repository.getPlugins())) {
                Log.add(
                        "[PluginUpdate] vBook Repository format detected, plugins: "
                                + repository.getPlugins().size());
                checkUpdateVBook(repository);
                return;
            }

            // Fall back to legacy array format
            Log.add("[PluginUpdate] Trying legacy array format...");
            JSONArray objArr = new JSONArray(js);
            Log.add("[PluginUpdate] Legacy format detected, plugins: " + objArr.length());
            checkUpdateLegacy(objArr);

        } catch (Exception e) {
            Log.add("[PluginUpdate] Error checking update: " + e.getMessage());
        }
    }

    /** Check updates for vBook format plugins. */
    private void checkUpdateVBook(Repository repository) {
        int updateCount = 0;
        int downloadCount = 0;

        for (Repository.Plugin pluginMeta : repository.getPlugins()) {

            try {
                var existingPlugin = PluginManager.getManager().get(pluginMeta.getPath());

                if (existingPlugin != null) {
                    // Update existing plugin
                    if (pluginMeta.getVersion() > existingPlugin.getVersion()) {
                        Log.add(
                                "[PluginUpdate] UPDATE: "
                                        + existingPlugin.getName()
                                        + " v"
                                        + existingPlugin.getVersion()
                                        + " -> v"
                                        + pluginMeta.getVersion());
                        updateCount++;
                    } else {
                        Log.add(
                                "[PluginUpdate] UP-TO-DATE: "
                                        + existingPlugin.getName()
                                        + " v"
                                        + existingPlugin.getVersion());
                    }
                }

            } catch (Exception e) {
                Log.add(
                        "[PluginUpdate] Error processing vBook plugin "
                                + pluginMeta.getName()
                                + ": "
                                + e.getMessage());
            }
        }

        Log.add(
                "[PluginUpdate] vBook update complete: "
                        + updateCount
                        + " updates, "
                        + downloadCount
                        + " new plugins");
    }

    /** Check updates for legacy format plugins. */
    private void checkUpdateLegacy(JSONArray plugins) {
        int updateCount = 0;
        int downloadCount = 0;

        for (int i = 0; i < plugins.length(); i++) {
            JSONObject obj = plugins.getJSONObject(i);
            String pluginUuid = obj.getString("uuid");
            String pluginName = obj.getString("name");
            double remoteVersion = obj.getDouble("version");

            try {
                boolean found = false;
                for (PluginEntity existingPlugin : PluginManager.getManager().list()) {
                    if (pluginUuid.equals(existingPlugin.getUuid())) {
                        found = true;
                        if (remoteVersion > existingPlugin.getVersion()) {
                            Log.add(
                                    "[PluginUpdate] UPDATE: "
                                            + existingPlugin.getName()
                                            + " v"
                                            + existingPlugin.getVersion()
                                            + " -> v"
                                            + remoteVersion);
                            updateLegacyPlugin(obj, existingPlugin);
                            updateCount++;
                        } else {
                            Log.add(
                                    "[PluginUpdate] UP-TO-DATE: "
                                            + existingPlugin.getName()
                                            + " v"
                                            + existingPlugin.getVersion());
                        }
                        break;
                    }
                }

                if (!found) {
                    Log.add("[PluginUpdate] NEW PLUGIN: " + pluginName + " v" + remoteVersion);
                    downloadLegacyPlugin(obj, pluginName, pluginUuid);
                    downloadCount++;
                }

            } catch (Exception e) {
                Log.add(
                        "[PluginUpdate] Error processing legacy plugin "
                                + pluginName
                                + ": "
                                + e.getMessage());
            }
        }

        Log.add(
                "[PluginUpdate] Legacy update complete: "
                        + updateCount
                        + " updates, "
                        + downloadCount
                        + " new plugins");
    }

    private void updateLegacyPlugin(JSONObject obj, PluginEntity existingPlugin) {
        try {
            String path =
                    AppUtils.curDir + "/tools/plugins/" + existingPlugin.getUuid() + ".plugin";
            String json = Http.request(obj.getString("url")).string();

            PluginEntity entity = gson.fromJson(json, PluginEntity.class);
            entity.setChecked(true);
            FileUtils.string2file(gson.toJson(entity), path);
            existingPlugin.apply(entity);
            PluginManager.notifyPluginsChanged();

            Log.add(
                    "[PluginUpdate] ✓ Updated: "
                            + existingPlugin.getName()
                            + " v"
                            + existingPlugin.getVersion());
            showToast(
                    "Đã update plugin "
                            + existingPlugin.getName()
                            + " v"
                            + existingPlugin.getVersion());

        } catch (Exception e) {
            Log.add(
                    "[PluginUpdate] ✗ Update failed for "
                            + existingPlugin.getName()
                            + ": "
                            + e.getMessage());
        }
    }

    private void downloadLegacyPlugin(JSONObject obj, String pluginName, String pluginUuid) {
        try {
            String path = AppUtils.curDir + "/tools/plugins/" + pluginUuid + ".plugin";
            Log.add("[PluginUpdate] Downloading: " + pluginName + " from " + obj.getString("url"));

            String json = Http.request(obj.getString("url")).string();
            PluginEntity entity = gson.fromJson(json, PluginEntity.class);
            entity.setChecked(true);
            FileUtils.string2file(gson.toJson(entity), path);
            PluginManager.getManager().add(path);

            Log.add("[PluginUpdate] ✓ Downloaded: " + pluginName);
            showToast("Đã tải xuống plugin " + pluginName);

        } catch (Exception e) {
            Log.add("[PluginUpdate] ✗ Download failed for " + pluginName + ": " + e.getMessage());
        }
    }

    private void showToast(String message) {
        SwingUtilities.invokeLater(() -> Toast.Build().content(message).time(3000).open());
    }
}

File: engine/src/commonMain/kotlin/dev/haipham22/leechtext/plugin/PluginUpdate.kt
package dev.haipham22.leechtext.plugin

import dev.haipham22.leechtext.entities.PluginEntity
import dev.haipham22.leechtext.log.EngineLogger
import dev.haipham22.leechtext.models.Repository
import dev.haipham22.leechtext.plugin.js.api.Http
import dev.haipham22.leechtext.plugin.vbook.VBookRepositoryClient
import dev.haipham22.leechtext.util.EnginePaths
import dev.haipham22.leechtext.util.LeechJson
import dev.haipham22.leechtext.util.writeTo
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject

/**
 * Check update plugin từ repository (port từ plugin/PluginUpdate.java; P5.2c common:
 * org.json → kotlinx.serialization, File → okio path string).
 */
@Suppress("LongParameterList") // deps engine DI — gộp thành config object chỉ thêm lớp vô ích
class PluginUpdate(
    private val log: EngineLogger,
    private val pluginManager: PluginManager,
    private val repositoryManager: RepositoryManager,
) {
    /**
     * Check update mọi repo enabled — chạy song song (thay legacy
     * Executors.newFixedThreadPool(MAX_CONN)).
     */
    suspend fun checkUpdate() {
        val repos = repositoryManager.repositoryList()
        if (repos.isEmpty()) return

        coroutineScope {
            repos
                .filter { it.isEnabled }
                .map { repo ->
                    async {
                        try {
                            checkUpdate(repo.link ?: return@async)
                        } catch (e: Exception) {
                            log.add(e)
                        }
                    }
                }.awaitAll()
        }
    }

    fun checkUpdate(repositoryLink: String?) {
        try {
            val js = VBookRepositoryClient.fetchRepoJson(repositoryLink ?: return, log)
            if (js.isEmpty()) {
                log.add("[PluginUpdate] Empty or null response from repository")
                return
            }

            // Try vBook Repository format trước — Gson trả null khi response là array
            // (legacy format); kotlinx throw nên bọc runCatching giữ fallback
            val repository = runCatching { LeechJson.decodeFromString<Repository>(js) }.getOrNull()
            val repoPlugins = repository?.plugins
            if (!repoPlugins.isNullOrEmpty()) {
                log.add(
                    "[PluginUpdate] vBook Repository format detected, plugins: " +
                        repoPlugins.size,
                )
                checkUpdateVBook(repository)
                return
            }

            // Fallback legacy array format
            log.add("[PluginUpdate] Trying legacy array format...")
            val objArr = LeechJson.parseToJsonElement(js).jsonArray
            log.add("[PluginUpdate] Legacy format detected, plugins: ${objArr.size}")
            checkUpdateLegacy(objArr)
        } catch (e: Exception) {
            log.add("[PluginUpdate] Error checking update: ${e.message}")
        }
    }

    /** Check update cho plugin vBook format. */
    private fun checkUpdateVBook(repository: Repository) {
        for (pluginMeta in repository.plugins ?: return) {
            try {
                val existingPlugin = pluginManager.get(pluginMeta.path ?: continue) ?: continue

                // pluginMeta.version là thang int repo (13 = 1.3) — chuẩn hóa /10
                // như PluginScreen; raw 13 > 1.3 → báo UPDATE ma vĩnh viễn (260905)
                val repoVersion = pluginMeta.version / 10.0
                if (repoVersion > (existingPlugin.version ?: 0.0)) {
                    log.add(
                        "[PluginUpdate] UPDATE: ${existingPlugin.name} " +
                            "v${existingPlugin.version} -> v$repoVersion",
                    )
                } else {
                    log.add(
                        "[PluginUpdate] UP-TO-DATE: ${existingPlugin.name} " +
                            "v${existingPlugin.version}",
                    )
                }
            } catch (e: Exception) {
                log.add(
                    "[PluginUpdate] Error processing vBook plugin ${pluginMeta.name}: " +
                        e.message,
                )
            }
        }
    }

    /** Check update cho plugin legacy format. */
    private fun checkUpdateLegacy(plugins: JsonArray) {
        var updateCount = 0
        var downloadCount = 0

        for (pluginJson in plugins) {
            val meta = parseLegacyPluginMeta(pluginJson) ?: continue
            try {
                val existingPlugin = pluginManager.list().firstOrNull { meta.uuid == it.uuid }
                if (existingPlugin == null) {
                    log.add("[PluginUpdate] NEW PLUGIN: ${meta.name} v${meta.remoteVersion}")
                    downloadLegacyPlugin(meta.obj, meta.name, meta.uuid)
                    downloadCount++
                } else {
                    updateCount += updateLegacyIfNewer(meta.obj, existingPlugin, meta.remoteVersion)
                }
            } catch (e: Exception) {
                log.add("[PluginUpdate] Error processing legacy plugin ${meta.name}: ${e.message}")
            }
        }

        log.add("[PluginUpdate] Legacy update complete: $updateCount updates, $downloadCount new plugins")
    }

    private data class LegacyPluginMeta(
        val obj: JsonObject,
        val uuid: String,
        val name: String,
        val remoteVersion: Double,
    )

    /** Parse metadata 1 plugin legacy; null khi JSON hỏng hoặc thiếu uuid/name/version. */
    private fun parseLegacyPluginMeta(pluginJson: JsonElement): LegacyPluginMeta? {
        val obj = runCatching { pluginJson.jsonObject }.getOrNull() ?: return null
        val uuid = obj.str("uuid")
        val name = obj.str("name")
        val remoteVersion = (obj["version"] as? JsonPrimitive)?.doubleOrNull
        return if (uuid != null && name != null && remoteVersion != null) {
            LegacyPluginMeta(obj, uuid, name, remoteVersion)
        } else {
            null
        }
    }

    /** UPDATE log + ghi đè file khi remote mới hơn; UP-TO-DATE log khi không. Trả 1 nếu đã update. */
    private fun updateLegacyIfNewer(
        obj: JsonObject,
        existingPlugin: PluginEntity,
        remoteVersion: Double,
    ): Int {
        if (remoteVersion <= (existingPlugin.version ?: 0.0)) {
            log.add(
                "[PluginUpdate] UP-TO-DATE: ${existingPlugin.name} " +
                    "v${existingPlugin.version}",
            )
            return 0
        }
        log.add(
            "[PluginUpdate] UPDATE: ${existingPlugin.name} " +
                "v${existingPlugin.version} -> v$remoteVersion",
        )
        updateLegacyPlugin(obj, existingPlugin)
        return 1
    }

    private fun updateLegacyPlugin(
        obj: JsonObject,
        existingPlugin: PluginEntity,
    ) {
        try {
            val path = "${pluginsDir()}/${existingPlugin.uuid}.plugin"
            val json = Http(log).request(obj.str("url") ?: return).string()

            val entity = LeechJson.decodeFromString<PluginEntity>(json)
            entity.checked = true
            LeechJson.encodeToString(entity).writeTo(path, log = log)
            existingPlugin.apply(entity)
            pluginManager.notifyPluginsChanged()

            log.add("[PluginUpdate] ✓ Updated: ${existingPlugin.name} v${existingPlugin.version}")
            // ponytail: Toast Swing bỏ — thay bằng log; thêm notification khi có UI layer
            log.add("Đã update plugin ${existingPlugin.name} v${existingPlugin.version}")
        } catch (e: Exception) {
            log.add("[PluginUpdate] ✗ Update failed for ${existingPlugin.name}: ${e.message}")
        }
    }

    private fun downloadLegacyPlugin(
        obj: JsonObject,
        pluginName: String,
        pluginUuid: String,
    ) {
        try {
            val path = "${pluginsDir()}/$pluginUuid.plugin"
            log.add("[PluginUpdate] Downloading: $pluginName from ${obj.str("url")}")

            val json = Http(log).request(obj.str("url") ?: return).string()
            val entity = LeechJson.decodeFromString<PluginEntity>(json)
            entity.checked = true
            LeechJson.encodeToString(entity).writeTo(path, log = log)
            pluginManager.add(path)

            log.add("[PluginUpdate] ✓ Downloaded: $pluginName")
            log.add("Đã tải xuống plugin $pluginName")
        } catch (e: Exception) {
            log.add("[PluginUpdate] ✗ Download failed for $pluginName: ${e.message}")
        }
    }

    private fun pluginsDir(): String = (EnginePaths.dataDir / "tools" / "plugins").toString()
}

private fun JsonObject.str(name: String): String? = (this[name] as? JsonPrimitive)?.takeIf { it !is kotlinx.serialization.json.JsonNull }?.content
